### DISCUSSION TITLE
using-a-culture-constraint-and-catching-404s-with-the-url-culture-provider/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#182] [Announcements]
# using-a-culture-constraint-and-catching-404s-with-the-url-culture-provider/

In this post, I'll show how we can create a custom route constraint to help prevent invalid route matching, and add additional routes to catch 404 errors.

https://andrewlock.net/using-a-culture-constraint-and-catching-404s-with-the-url-culture-provider/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at January 26 2017, 04:05, in reply to [wc.matteo](https://disqus.com/by/wcmatteo/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/182#discussioncomment-6024654)</em>

---

<p>Thanks, I'll update the post:) </p>

[a comment by an automated account is not shown]

**Reply by MSingh-13** [reply]
@jalza. I faced the same issue with the MapGet() route on .NET 7. This seems to be due to a fix applied in .NET 5 which is explained in https://github.com/dotnet/aspnetcore/issues/29594. From .NET 5 onwards the catch all route (*path) has an order of 0 while conventional route (MapControllerRoute) has an order of 1. So with the above example the MapGet route with {*path} gets matched first and executed. The fix is to use MapFallback instead of MapGet, which will give the route an order of int.MaxValue and make it the last option.

Also now the routes can be defined by calling methods on WebApplication.

Here is my verison of the routes in .NET 7

```
static void LocalisationRoutes(WebApplication app)
{
    //default route
    app.MapControllerRoute(
                name: "default",
                pattern: "{culture:culturecode}/{controller=home}/{action=index}/{id?}");

    //fallback route if default route does not match
    app.MapFallback("{culture:culturecode}/{*path}", ctx =>
    {
        ctx.Response.StatusCode = 404;
        return Task.CompletedTask;
    });

    //fallback route to add culture to url
    app.MapFallback("{*path}", ctx =>
    {
        var localizationOptions = app.Services.GetService<IOptions<RequestLocalizationOptions>>().Value;
        var defaultCulture = localizationOptions.DefaultRequestCulture.Culture.Name;
        var path = ctx.GetRouteValue("path") ?? string.Empty;
        var culturedPath = $"/{defaultCulture}{ctx.Request.GetEncodedPathAndQuery()}";
        ctx.Response.Redirect(culturedPath);
        return Task.CompletedTask;
    });
}
```

**Reply by MSingh-13** [reply]
Below is an updated version of LocalisationRoutes to cover the cases where the application is not at the root of the website.

```
public static void AddLocalisationRoutes(WebApplication app)
{
    //default route
    app.MapControllerRoute(
                name: "default",
                pattern: "{culture:culturecode}/{controller=home}/{action=index}/{id?}");

    //fallback route if default route does not match
    app.MapFallback("{culture:culturecode}/{*path}", ctx =>
    {
        ctx.Response.StatusCode = 404;
        return Task.CompletedTask;
    });

    //fallback route to add culture to url
    app.MapFallback("{*path}", ctx =>
    {
        var localizationOptions = app.Services.GetRequiredService<IOptions<RequestLocalizationOptions>>().Value;
        var defaultCulture = localizationOptions.DefaultRequestCulture.Culture.Name;
        var path = ctx.GetRouteValue("path") ?? string.Empty;
        var queryString = ctx.Request.QueryString;
        var culturedPath = $"{ctx.Request.PathBase}/{defaultCulture}/{path}{queryString}";
        ctx.Response.Redirect(culturedPath);
        return Task.CompletedTask;
    });
}
```
